### ISSUE TITLE
`ToPrimitive` bound on `Iterator` impl for `Range` and related traits.

### ISSUE THREAD

**Issue by alercah** [rust-num/num-iter#13]
See [this line](https://docs.rs/num-iter/0.1.39/src/num_iter/lib.rs.html#87) which references rust-lang/rust#10414, now closed. The modern `Range` trait in the stdlib uses an unstable [`Step`](https://doc.rust-lang.org/std/iter/trait.Step.html) trait to generate its `size_hint`; perhaps something similar would work for this crate?

**Comment by cuviper** [comment]
Well, `core` can get away with that because they only support iterating `Range` with the primitive integer types, which can privately implement `Step`. For `num-iter`, we would need a public (stable) trait that external users can implement to make this `Range` work for their types.

**Comment by alercah** [comment]
Yes, agreed, it would need to be stable. But that seems more flexible than requiring `ToPrimitive`.
